### PR TITLE
Trampoline routing proposal

### PR DESCRIPTION
**Pull request by t-bast**
This PR highlights proposed changes to enable trampoline routing (once #593 and #603 are accepted).

Everything is currently grouped in a single file to make the initial review process easier. Once we start converging on the proposal, the final changes should probably be split between bolts 2, 4 and 7 (unless you feel this should be a new bolt) and the wording should be in formal spec format (with more details on some sections).

### CODE REVIEW

**Comment by cdecker** [comment]
FWIW I think it is a great idea to have an overall design doc like this that collects all the details of the incremental change in one place. We agreed on having a "living" spec, but I still like having a well-written and easy to follow bundle that explains the rationale more in depth :+1:

File: wip-trampoline.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+## Packet Structure
+
+### Trampoline Onion
+
+We define a fixed-size trampoline onion packet that can be nested inside onion
+packets. We define `N` to be the maximum number of trampoline nodes (excluding
+the sender node and the destination node).
+
+The trampoline packet always contains `2*N*FRAME_SIZE + 66` bytes.
```
Comment (cdecker) [inline wip-trampoline.md]: I'm not clear where the `2*FRAME_SIZE` per hop comes from.

File: wip-trampoline.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+Note: we use the same onion format as the standard onion which wastes frames,
+but that theoretically allows another level of nesting (mind blown).
+However not really usable given the space constraints...
+
+TODO: maybe this should be version 1 of the overall onion version namespace?
+
+TODO: maybe we can do something more clever with TLV? At the cost of a second
+onion parser (trampoline onion parser)? Use the top-level version in the parser
+to discriminate?
```
Comment (cdecker) [inline wip-trampoline.md]: If we stick with 65 byte frames we can reuse the onion processing from the outer onion.

File: wip-trampoline.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+   * [`32`:`HMAC`]
+
+For this specification, some restrictions apply:
+
+* `realm` MUST be `0x01` to indicate that the payload contains a TLV stream
+and `num_frames` MUST correctly reflect the number of frames used.
+* the TLV stream MUST contain the `type` `0x01` (`next_trampoline_hop`)
+* `next_trampoline_hop`'s length MUST be `33+8+4=45` bytes
+* the TLV stream MAY contain other optional types
+* `padding` MUST use only `0x00` bytes
```
Comment (cdecker) [inline wip-trampoline.md]: I'd prefer sticking with a raw TLV on the top-level that is easier to change later, rather than rebuilding the legacy `hop_data_v0` format in this new layer.

File: wip-trampoline.md (hunk 1 above)
Comment (cdecker) [reply wip-trampoline.md]: I see, it comes from the requirement of 45 bytes of payload, however if we just describe this as a TLV payload the rest resolves itself. That is if we just say we have `N` frames available, and allow any trampoline hop to use as many as they want we're not imposing this limitation right away.

Who knows, we might come up with a more concise way of addressing a node, and the amount to forward and we'd suddenly fit into a single frame.

Long story short: I think it's a mistake to hard-code the `2*FRAME_SIZE` here.

File: wip-trampoline.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+1. type: `onion_packet`
+2. data:
+   * [`1`:`version`]
+   * [`33`:`public_key`]
+   * [`n1*FRAME_SIZE`:`hop_payload`]
+   * [`n2*FRAME_SIZE`:`hop_payload`]
+   * ...
+   * [`2*(N+1)*FRAME_SIZE`:`trampoline_onion_packet_payload`]
+   * `filler`
+   * [`32`:`hmac`]
```
Comment (cdecker) [inline wip-trampoline.md]: So in this example `hop_payload`s of `n1`, `n2` and so forth are the path to get to the trampoline node, and then the `trampoline_onion_packet_payload` is used to find a continuation at the trampoline, right?

I don't understand the rationale of `2*(N+1)*FRAME_SIZE` which is bigger than the payload we have at our disposal.

File: wip-trampoline.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+The structure of the last payload is:
+
+1. type: `trampoline_onion_packet_payload`
+2. data:
+   * [`1`:`realm`]
+   * [`1`:`type`]
+   * [`3`:`length`]
+   * [`2*N*FRAME_SIZE + 66`:`trampoline_onion_packet`]
+   * `optional TLV can be added here`
+   * `padding`
```
Comment (cdecker) [inline wip-trampoline.md]: Why not just define a TLV type and make it a completely normal TLV type-length-value pair? That way we could even give additional instructions to the trampoline, instead of having an optional TLV at the end.

File: wip-trampoline.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+improved.
+
+Upon receiving a packet with type `trampoline_onion_packet_payload`:
+
+* The receiving node is the destination for this v0 onion packet
+* It extracts the `trampoline_onion_packet` and is able to unwrap one layer to
+discover the next trampoline node
+* It creates a new v0 onion packet to the next trampoline node
+* It puts the unwrapped `trampoline_onion_packet` in the payload for the last
+hop
```
Comment (cdecker) [inline wip-trampoline.md]: Notice that we also need a downgrade option: while we may want to rely on trampolines to forward a payment through the network and each of the trampolines just receives an inner onion it knows how to process, the recipient may not support trampolines at all. Hence the last trampoline may be asked to downgrade to a regular `hop_payload` so that the onion arriving at the recipient is just a classical onion.

Comment (cdecker) [reply wip-trampoline.md]: While this seems to be addressed in the example below I can't find a description of how this should really work.

File: wip-trampoline.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+P -> R1 -> R2 -> R3 -> M
+
+R1, R2 and R3 need to have the `option_trampoline` feature flag set.
+Setting this flag implies that you fully comply with the trampoline
+specification.
+
+M doesn't need to understand trampoline nor multi-frame: the last trampoline
+hop will use a v0 onion to route the payment to M.
+
+M will usually include a few "close" trampoline nodes in its invoices to help
```
Comment (cdecker) [inline wip-trampoline.md]: Well, if `M` doesn't know about trampoline, how is he supposed to find close-by trampoline nodes?

File: wip-trampoline.md [hunk 8]
```diff
@@ -25,6 +25,15 @@ but that theoretically allows another level of nesting (mind blown).
...
 TODO: maybe this should be version 1 of the overall onion version namespace?
+It can then also be used as a top-level onion (smaller than the v0 onion).
+It's like the v0 onion but lets each hop choose how they route to the next hop.
+If they have a direct channel they directly send the unwrapped v1 onion to the
+next hop.
+If they don't they wrap the v1 onion inside a v0 to get to the next hop via
+intermediate hops.
+Even though the onion packet is smaller, you don't know if previous hops used
+the direct route or went through v0 wrappings, so you have even greater doubts
+about your position in the route than with only v0.
```
Comment (cdecker) [inline wip-trampoline.md]: I disagree with the fact that v1 should be a trampoline onion. The onion construction we have is versatile enough to be reused as is for the trampoline onion.

The chances of that happening is virtually 0 and we shouldn't be optimizing for this case, and you are basically giving away to an observer that this was a trampoline onion and not a routing onion, by just observing a smaller packet go over the encrypted connection.

File: wip-trampoline.md [hunk 9]
```diff
@@ -1,60 +1,40 @@
...
-
-TODO: maybe we can do something more clever with TLV? At the cost of a second
-onion parser (trampoline onion parser)? Use the top-level version in the parser
-to discriminate?
-
-TODO: version maybe redundant with the TLV type that will include this?
-But keeping it allows using this packet to be used as a top-level onion packet
-where we basically let each hop choose which channel to use to route to the
-next hop. Is this useful?
+This results in a `706`-bytes fixed-size packet.
```
Comment (cdecker) [inline wip-trampoline.md]: Just as a side-note: this leaves us with 594 bytes for the remaining hops, or a maximum routing distance of 9 between two consecutive trampolines.

Comment (cdecker) [reply wip-trampoline.md]: It also means that the maximum route length that is actually enforced by the protocol is `9 * 8 = 72`, which gives us a 72x multiplier for griefing attacks...

File: wip-trampoline.md [hunk 10]
```diff
@@ -63,36 +43,67 @@ Each `trampoline_hop_payload` has the following structure:
 1. type: `trampoline_hop_payload`
 2. data:
    * [`1`:`num_frames_and_realm`]
-   * [`1`:`type`]
-   * [`1`:`length`]
+   * [`47`:`next_trampoline_node`]
```
Comment (cdecker) [inline wip-trampoline.md]: Like mentioned before `next_trampoline_node` should itself be expressed as TLV-pairs in a single TLV, in case we want to switch out the way we address the nodes, allowing us to reduce the footprint.

File: wip-trampoline.md [hunk 11]
```diff
@@ -63,36 +43,67 @@ Each `trampoline_hop_payload` has the following structure:
...
-   * `padding`
-   * [`32`:`HMAC`]
 
-For this specification, some restrictions apply:
+#### Requirements
+
+A sending node:
+
+* MUST set the `realm` part of `num_frames_and_realm` to `0x01` to indicate
+that the payload contains a TLV stream
```
Comment (cdecker) [inline wip-trampoline.md]: This itself implies that the entire payload is just one TLV.

File: wip-trampoline.md [hunk 12]
```diff
@@ -187,12 +187,97 @@ themselves.
...
+2. data:
+   * [`64`:`signature`]
+   * [`32`:`chain_hash`]
+   * [`33`:`node_id`]
+   * [`4`:`timestamp`]
+   * [`2`:`cltv_expiry_delta`]
+   * [`8`:`htlc_minimum_msat`]
+   * [`8`:`htlc_maximum_msat`]
+   * [`4`:`fee_base_msat`]
+   * [`4`:`fee_proportional_millionths`]
```
Comment (cdecker) [inline wip-trampoline.md]: Unless this message is also gated by the existence of a channel with the node as an endpoint it becomes a really cheap DoS attack vector. It's also hard for a node with a partial network view to even verify that there is a channel matching this node.

File: wip-trampoline.md [hunk 13]
```diff
@@ -187,12 +187,97 @@ themselves.
...
+
+TODO: use a TLV stream for `node_update`.
+TODO: should use an odd type (261?) because it's safe to ignore if unsupported?
```
Comment (cdecker) [inline wip-trampoline.md]: Definitely :+1:

File: wip-trampoline.md [hunk 14]
```diff
@@ -187,6 +187,9 @@ themselves.
 
 This grows the maximum route length to `9 * 8 = 72` total hops.
 
+The same kind of mechanism can be used to easily unlock rendezvous routing by
+simply defining a new TLV type (that will extend the `trampoline_onion` type).
```
Comment (cdecker) [inline wip-trampoline.md]: Uhm, actually the rendez-vous routing as proposed [here](https://github.com/lightningnetwork/lightning-rfc/wiki/Rendez-vous-mechanism-on-top-of-Sphinx) works fine without the need to extend the `trampoline_onion`. It just needs a `switch_ephemeral_key` TLV entry and we're good to go :wink:

File: wip-trampoline.md [hunk 15]
```diff
@@ -229,33 +229,48 @@ TODO: use a TLV stream for `node_update`.
...
+If a node is not willing to relay trampoline payments, it will simply never
+send a `node_update`. This lets the network discriminate nodes that are able to
+relay trampoline payments from nodes that use it but simply as sender or
+destination.
 
 ### Fees and CLTV estimation
 
-TODO: detail heuristics that can be used by trampoline nodes to estimate cltv
-and fees.
+Trampoline nodes need to estimate a `cltv_expiry_delta` and `fee` that would
```
Comment (cdecker) [inline wip-trampoline.md]: Trampoline nodes should be encouraged to not be constrained and keep a full view of the topology.

File: wip-trampoline.md [hunk 16]
```diff
@@ -229,33 +229,48 @@ TODO: use a TLV stream for `node_update`.
...
+allow them to route to any other trampoline node while being competitive with
+other nodes. This is a great opportunity to incentivize nodes to open channels
+between each other to minimize the cost of trampoline hops. This is also a
+great opportunity for nodes to implement smart fee estimation algorithms as
+a competitive advantage.
```
Comment (cdecker) [inline wip-trampoline.md]: Is it? I would expect the trampoline to just do trial-and-error with the maximum fee allocated to themselves just like the original sender would. That's not really estimating fees at all. They even get told how much funds MUST arrive at the next trampoline.

File: wip-trampoline.md (hunk 1 above)
Comment (t-bast) [reply wip-trampoline.md]: This was an old commit and was removed afterwards.
Good news is that with yesterday's proposal during the IRC meeting, we're getting rid of `FRAME_SIZE` entirely ;)

File: wip-trampoline.md (hunk 2 above)
Comment (t-bast) [reply wip-trampoline.md]: Same, this was an old commit. I plan on removing `FRAME_SIZE` entirely, so it's going to be a lot nicer. I'll also squash all the commits together to hide the old commits.

File: wip-trampoline.md (hunk 3 above)
Comment (t-bast) [reply wip-trampoline.md]: I agree, I'll clarify this while removing the fixed frames entirely.

File: wip-trampoline.md (hunk 4 above)
Comment (t-bast) [reply wip-trampoline.md]: > So in this example hop_payloads of n1, n2 and so forth are the path to get to the trampoline node, and then the trampoline_onion_packet_payload is used to find a continuation at the trampoline, right?

Exactly.

> I don't understand the rationale of 2*(N+1)*FRAME_SIZE which is bigger than the payload we have at our disposal.

I think the confusion comes from the distinction between `FRAME_SIZE` and `FRAME_SIZE_V1`.
Removing fixed sizes entirely will make the proposal easier to understand.

File: wip-trampoline.md (hunk 5 above)
Comment (t-bast) [reply wip-trampoline.md]: I'll investigate that, it's a bit of a trade-off between flexibility and space.
Thanks for the suggestion!

File: wip-trampoline.md (hunk 6 above)
Comment (t-bast) [reply wip-trampoline.md]: Very good point, for now I assumed that the destination would need to handle trampoline (and ideally provide an invoice with hints about close trampoline nodes).
However you're right that we should make this work for legacy payee nodes. I'll figure something out.

File: wip-trampoline.md (hunk 7 above)
Comment (t-bast) [reply wip-trampoline.md]: Magic ¯\_(ツ)_/¯
This was an old commit, I fixed this by explicitly stating that M needed to understand trampoline, but in fact I must make this work for payees that don't support it. I'll work on that.

File: wip-trampoline.md (hunk 8 above)
Comment (t-bast) [reply wip-trampoline.md]: This is an important point, this is a fundamental design choice. Let's discuss it in the global conversation for visibility (your comment was made on an old commit).

File: wip-trampoline.md (hunk 10 above)
Comment (t-bast) [reply wip-trampoline.md]: I'll explore that option.

File: wip-trampoline.md (hunk 12 above)
Comment (t-bast) [reply wip-trampoline.md]: Nodes will apply the same filters for this message as what they do for `channel_update`s.
A node producing the `node_update` can only send it to peers with whom it has a channel and can't send `node_update`s more than once every X seconds.
Most nodes that support the `node_update` message will be routing/trampoline nodes and will have the full network view, so they can drop messages from nodes that don't have any channel.

You have a very good point though for the constrained nodes that have only a view of their neighbourhood (let's call one of those Alice). If one of Alice's neighbours is malicious and forwards `node_update`s for nodes that don't have any channel, Alice can't verify that and might include that node in her trampoline routes which is an issue. However, Alice will regularly change her list of trampoline nodes based on randomness and reliability: payment failure will trigger the eviction of the malicious node.

That needs to be detailed more thoroughly though.

File: wip-trampoline.md (hunk 14 above)
Comment (t-bast) [reply wip-trampoline.md]: That `switch_ephemeral_key` is a cute trick, good job ;)
With trampoline we're able to do rendezvous without switching the ephemeral key. But your proposal with `switch_ephemeral_key` works great too and might require less space.

File: wip-trampoline.md (hunk 15 above)
Comment (t-bast) [reply wip-trampoline.md]: Yes definitely. I think this incentivizes them to do it because it's the best way for them to be more economically viable.

File: wip-trampoline.md (hunk 16 above)
Comment (t-bast) [reply wip-trampoline.md]: > They even get told how much funds MUST arrive at the next trampoline.

But that will often be based on the fee they advertised (via their estimation algorithm) ;)

> I would expect the trampoline to just do trial-and-error

They could, but there's a risk they will end up locking up funds if one of the nodes in the route is unresponsive and they have to wait until the timeout. It's hard to predict because I think the heuristics will depend on the reliability of the overall network, but I think both options could make sense.

**Comment by t-bast** [comment]
Thanks for all the feedback @cdecker, really appreciate it (especially since this is a somewhat dense early draft).

> We agreed on having a "living" spec, but I still like having a well-written and easy to follow bundle that explains the rationale more in depth

I think we could use the wiki for that (like what you did for rendezvous). Once the trampoline proposal is integrated to the spec (in spec format, probably split between multiple bolt pages), a wiki page to explain the rationale and give an overall view of how all pieces come together is going to be useful. What do you think?

To sum up your comments for other readers, here are the important things I need to fix (let me know if there is something I missed or misunderstood):

* The fixed-size framing is messy and makes the proposal hard to follow: thanks to yesterday's variable-length frames proposal, I'll remove fixed frames entirely and it should be a lot simpler
* I should use finer-grained TLVs
* The payee doesn't need to understand trampoline (last hop needs to switch to v0 onion if needed)
* I need to clarify the restrictions on `node_update`s to make sure this doesn't introduce an attack vector

> you are basically giving away to an observer that this was a trampoline onion and not a routing onion, by just observing a smaller packet

If I understand this comment correctly, you'd like me to remove the proposal to define onion v1 as a 706-bytes packet. Instead I could only define the trampoline onion as a 706-bytes TLV that lives inside a v0 onion (and we would keep only 1366-bytes sized onion packets). Would that work for you?

Side-note: it looks like old commits made it harder to review - do you think I should squash everything to start from a clean slate and not confuse reviewers?

File: wip-trampoline.md [hunk 17]
```diff
@@ -0,0 +1,840 @@
...
+Each `trampoline_hop_payload` has the following structure:
+
+1. type: `trampoline_hop_payload`
+2. data:
+   * [`var_int`:`length`]
+   * [`length`:`tlv_stream`]
+   * [`32`:`hmac`]
+
+The `tlv_stream` contains data necessary to route to the next trampoline hop.
+The following types MAY be used to encode that data:
```
Comment (t-bast) [inline wip-trampoline.md]: It feels like many TLV types will be useful accross many message types.
I'd like to have a centralized place where we could register known TLVs and their assigned `type` (like what we have in bolt #9). I think that all the TLVs I'm using here (except the `trampoline_onion`) will be useful to other messages and should be specified outside of this PR.

**Comment by t-bast** [comment]
Closing in favor of #654
